lib/std/crypto/bcrypt.zig

```diff
@@ -183,28 +200,73 @@ const Codec = struct {
         var i: usize = 0;
         var j: usize = 0;
         while (j < bin.len) {
-            const c1 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i]) orelse return error.InvalidEncoding);
-            const c2 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i + 1]) orelse return error.InvalidEncoding);
+            const c1 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i]) orelse return Error.InvalidEncoding);
+            const c2 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i + 1]) orelse return Error.InvalidEncoding);
             bin[j] = (c1 << 2) | ((c2 & 0x30) >> 4);
             j += 1;
             if (j >= bin.len) {
                 break;
             }
-            const c3 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i + 2]) orelse return error.InvalidEncoding);
+            const c3 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i + 2]) orelse return Error.InvalidEncoding);
             bin[j] = ((c2 & 0x0f) << 4) | ((c3 & 0x3c) >> 2);
             j += 1;
             if (j >= bin.len) {
                 break;
             }
-            const c4 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i + 3]) orelse return error.InvalidEncoding);
+            const c4 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i + 3]) orelse return Error.InvalidEncoding);
             bin[j] = ((c3 & 0x03) << 6) | c4;
             j += 1;
             i += 4;
         }
     }
 };
 
-fn strHashInternal(password: []const u8, rounds_log: u6, salt: [salt_length]u8) Error![hash_length]u8 {
+pub const Params = struct {
+    const Self = @This();
+
+    log_rounds: u6,
+
+    /// Public interface for PhcEncoding
+    pub fn fromPhcEncoding(it: *PhcParamsIterator) Error!Self {
+        var lr: ?u6 = null;
+        while (try it.next()) |param| {
+            if (mem.eql(u8, param.key, "lr")) {
+                lr = try param.decimal(u6);
+            } else {
+                return Error.InvalidEncoding;
+            }
+        }
+        return Self{ .log_rounds = lr orelse return Error.InvalidEncoding };
+    }
+
+    /// Public interface for PhcEncoding
+    pub fn toPhcEncoding(self: Self, out: []?PhcParamsIterator.Param) void {
+        var lr = PhcParamsIterator.Param.Value{};
+
+        const s = fmt.bufPrint(&lr.buf, "{d}", .{self.log_rounds}) catch unreachable;
+        lr.len = s.len;
+
+        out[0] = PhcParamsIterator.Param{ .key = "lr", .value = lr };
+    }
+};
+
+fn phcKdf(
+    allocator: *mem.Allocator,
+    derived_key: []u8,
+    password: []const u8,
+    salt: []const u8,
+    params: Params,
+) !void {
+    kdf(derived_key[0..derived_key_length], password, salt[0..salt_length].*, params);
+}
+
+/// Apply BCRYPT to generate a key from a password.
```

Comment:
Wait, what? bcrypt is not a KDF.

Reply:
Let's look.. 

[wikipedia bcrypt](https://en.wikipedia.org/wiki/Bcrypt) tells us that: `bcrypt is a password-hashing function designed by..`, if we click on [password-hashing function](https://en.wikipedia.org/wiki/Password-hashing_function) we get redirected to [Key derivation function](https://en.wikipedia.org/wiki/Key_derivation_function). Where we can read this: 

`In cryptography, a key derivation function (KDF) is a cryptographic hash function that derives one or more secret keys from a secret value such as a main key, a password, or a passphrase using a pseudorandom function."`
...
`The first[citation needed] deliberately slow (key stretching) password-based key derivation function was called "crypt" (or "crypt(3)" after its man page), and was invented by Robert Morris in 1978.`
...
`Various algorithms have been designed specifically for this purpose, including bcrypt, scrypt and, more recently, Lyra2 and Argon2 (the latter being the winner of the Password Hashing Competition).`

Reply:
bcrypt is a little bit different: it was only made to be a password hashing function.

bcrypt as a KDF generally refers to the [bcrypt-pbkdf](https://flak.tedunangst.com/post/bcrypt-pbkdf) function.